blog/2024-06-02-kqir-query-engine/index.md

```diff
@@ -0,0 +1,222 @@
+
+Subsequently, Kvrocks retrieves query results through iterative execution.
+
+### On-disk indexing
+
+Unlike Redis, which stores index data in memory, Kvrocks requires the construction of indexes on the disk.
+This means that for any field type (e.g. tag, numeric), we need an encoding to reduce such index to rocksdb key-values.
+
+Furthermore, we incrementally create indexes before and after JSON or HASH commands getting executed to guarantee that query results are in real-time.
+
+## Current status
+
+The KQIR functionality is currently available on the `unstable` branch, supporting commands like `FT.CREATE`, `FT.SEARCH`, and `FT.SEARCHSQL` (an extension for running SQL queries) to encourage user to test.
+
+However, as KQIR is still in early development, compatibility cannot be guaranteed and many features remain incomplete.
+Thus the upcoming release (version 2.9.0) will not include any KQIR component.
+
+### Supported field types
+
+Currently, we only support two field types: tag and numeric.
+
+Tag fields label each data record with multiple tags for filtering in queries.
+And numeric fields hold numerical data within double-precision floating-point ranges. They allow sorting and filtering by specific numerical ranges.
+
+In the future, we plan to expand support to include vector search and full-text search capabilities alongside other field types.
+
+### Transaction guarantees
+
+Currently, the transaction guarantee of KQIR is weak, which may lead to unexpected issues during use.
+
+[Another project in the Kvrocks community](https://github.com/apache/kvrocks/issues/2331) aims to enhance Kvrocks' transaction support by establishing a structured framework.
+We will leverage these efforts to uphold the ACID properties of KQIR and release an official version incorporating KQIR after that.
+
+### Limitation on IR optimizer
+
+Currently, KQIR needs to enhance the optimization implementation of plan operators such as `Sort`.
+
+Furthermore, KQIR does not currently utilize optimizations based on runtime statistics.
+Our future focus will be on integrating runtime statistics into the cost model for more precise index selection.
+
+### Relationship with other features
+
+KQIR integrates well with the [namespace](https://kvrocks.apache.org/docs/namespace) feature.
+Any index created is restricted to the current namespace and cannot be accessed in other namespaces, aligning with how other data is accessed within the namespace.
+
+Currently, KQIR cannot be enabled in the [cluster mode](https://kvrocks.apache.org/docs/cluster).
+
+### High experimental!
+
+The current implementation of KQIR is in its early experimental stage.
+As we do not guarantee compatibility and there may be unexpected errors, we advise users against using it in a production environment.
```

Comment:
```suggestion
We advise users to consider carefully when using KQIR functionalities in a production environment, as we do not guarantee compatibility, and there may be unexpected errors.
```

Reply:
We should encourage usage while keeping altering the instability. It's unreasonable to prevent users from using it.

Reply:
Got it.